src/state/slices/opportunitiesSlice/opportunitiesSlice.ts

```diff
@@ -75,35 +41,108 @@ export const opportunities = createSlice({
+        ...payload.byId,
       }
-      state[payload.type].ids = Array.from(new Set([...Object.keys(payload.metadata)]))
+      draftState[payload.type].ids = Array.from(new Set([...Object.keys(payload.byId)]))
+    },
+    // TODO: testme
+    upsertOpportunityAccounts: (
+      draftState,
+      { payload }: { payload: GetOpportunityUserDataOutput },
+    ) => {
+      draftState[payload.type].byAccountId = merge(
+        draftState[payload.type].byAccountId,
+        payload.byAccountId,
+      )
     },
     upsertUserStakingOpportunities: (
-      state,
+      draftState,
       { payload }: { payload: OpportunitiesState['userStaking']['byId'] },
     ) => {
-      state.userStaking.byId = {
-        ...state.staking.byId,
+      draftState.userStaking.byId = {
+        ...draftState.staking.byId,
         ...payload,
       }
-      state.userStaking.ids = Array.from(new Set([...Object.keys(payload)])) as UserStakingId[]
+      draftState.userStaking.ids = Array.from(new Set([...Object.keys(payload)])) as UserStakingId[]
     },
-
-    // upsertOpportunitiesData: (opportunitiesSliceDraft, { payload }: { payload: {} }) => {}, // TODO:
   },
 })
 
 export const opportunitiesApi = createApi({
   ...BASE_RTK_CREATE_API_CONFIG,
   reducerPath: 'opportunitiesApi',
   keepUnusedDataFor: 300,
-  endpoints: _build => ({
-    // TODO
+  endpoints: build => ({
+    getOpportunityMetadata: build.query<GetOpportunityMetadataOutput, GetOpportunityMetadataInput>({
+      queryFn: async ({ opportunityId, opportunityType, defiType }, { dispatch, getState }) => {
+        const resolver = getMetadataResolversByDefiProviderAndDefiType(
+          DefiProvider.FoxFarming,
+          defiType,
+        )
+        const resolved = await resolver({
+          opportunityId,
+          opportunityType,
+          reduxApi: { dispatch, getState },
+        })
+
+        dispatch(opportunities.actions.upsertOpportunityMetadata(resolved.data))
+
+        return { data: resolved.data }
+      },
+    }),
+    getOpportunityUserData: build.query<GetOpportunityUserDataOutput, GetOpportunityUserDataInput>({
+      queryFn: async (
+        { accountId, opportunityId, opportunityType, defiType },
+        { dispatch, getState },
+      ) => {
+        try {
+          const resolver = getUserDataResolversByDefiProviderAndDefiType(
+            DefiProvider.FoxFarming,
+            defiType,
+          )
+
+          if (!resolver) {
+            throw new Error(`resolver for ${DefiProvider.FoxFarming}::${defiType} not implemented`)
+          }
+
+          // TODO: This commit authors LP slice population only - for Fox staking we will want to assign this to a variable and actually use the data
+          // The reason for that is for EVM chains LPs, we just need to await this promise resolution - if this resolved, it means we have portfolio data
+          // If this throws, the RTK query is rejected and we never insert that AccountId into state
```

Comment:
This also implies that staking resolution will be slightly different, for the LP one, we literally only introspect portfolio data